--- a/sys/arm/broadcom/bcm2835/bcm2835_sdhci.c
+++ b/sys/arm/broadcom/bcm2835/bcm2835_sdhci.c
@@ -316,7 +316,7 @@ bcm_sdhci_attach(device_t dev)
 	sc->sc_slot.quirks = sc->conf->quirks;
 
 	sdhci_init_slot(dev, &sc->sc_slot, 0);
-	mmc_fdt_parse(dev, 0, &sc->sc_mmc_helper, &sc->sc_slot.host);
+	mmc_fdt_parse(dev, ofw_bus_get_node(dev), &sc->sc_mmc_helper, &sc->sc_slot.host);
 
 	sc->sc_dma_ch = bcm_dma_allocate(BCM_DMA_CH_ANY);
 	if (sc->sc_dma_ch == BCM_DMA_CH_INVALID)
